// Copyright (C) 2015 Thermo Fisher Scientific. All Rights Reserved.
/*
 * BbcUsage.h
 *
 * Contains the one function for all the help text available from the command-line.
 * Essentially to remove clutter from BbcMain.cpp, where it is included.
 *
 *  Created on: Nov 24, 2015
 *      Author: Guy Del Mistro
 */

#ifndef BBCUSAGE_H_
#define BBCUSAGE_H_

#include <string>
using namespace std;

static void Usage( string subcmd, bool verbose = false ) {
	if( subcmd.empty() ) {
		cout << "Usage:\n"
		"  bbctools <subcmd> [options] [files]\n\n"
		"Description:\n"
		"  Creates binary base coverage (BBC) files from binary alignment/map (BAM) files, recording just\n"
		"  forward and reverse base coverage at each position along a reference (genome) and whether each\n"
		"  was on target to any region, with an optional input BED file specified. Also provides tools to\n"
		"  view regions and analyze coverage statistics with various options. Performance of viewing BBC file\n"
		"  operations are highly optimized, especially when using BBC indexing files that may also be created.\n"
		"  Other options produce aligned read summary (text) files that are well suited to targeted (enriched)\n"
		"  sequencing techniques, such as PCR and ThermoFisher AmpliSeq and TargetSeq applications.\n\n"
		"Sub-commands:\n"
		"  create:   Create files (or stream to STDOUT) from input BAM or BBC file(s).\n"
		"  report:   Report a summary of base coverage for a given BBC file. Optionally output depth of coverage.\n"
		"  version:  Print current version of bbctools or a bbctools file format.\n"
		"  view:     View base coverage from a BBC file with output, range and binned coverage options.\n\n"
		"Type: 'bbctools <subcmd> -h (or --help) for (detailed) information on a particular sub-command.\n";
	} else if( subcmd == "create" ) {
		cout << "Usage:\n"
		"  bbctools create [options] (<BBC file> | <BAM file(s)>)\n\n";
		if( verbose ) {
			cout << "Description:\n"
			"  Creates one or more files from a BBC file or one or more BAM files as input, and/or prints\n"
			"  text output depending on the options selected. Some options may require BAM files as input.\n\n"
			"Primary file creation/output options:\n"
			"  -B,--bbc <file prefix>    File name (prefix) for BBC file and other bbctools binary file creation,\n"
			"      including BCI and CBC files if the -i and/or -c options are provided. A value of '-' may be\n"
			"      used to alternatively write the contents to STDOUT, which is also the default behavior for the\n"
			"      bbctools create and view commands where no options are provided. Refer to usage examples below.\n"
			"  -C,--covStats <file>      File name for coverage statistics output file, which is a tab-delimited\n"
			"      text file with calculated fields that depend on the --readType and other options specified.\n"
			"      Base and/or read coverage is provided per contig/chromosome, or for each target region if the\n"
			"      --regions (-R) option is provided. A value of '-' may be used to pipe this output to STDOUT.\n"
			"  -O,--readOrigin <file>    File name for read origin output file, which is a tab-delimited text file\n"
			"      with the assigned target and/or aligned length for every mapped read in the input BAM file.\n"
			"      The target is identified as the line number of the output --covStats file, or 0 if the read did\n"
			"      not overlap any target. (It is also the line number for targets defined in the input --regions\n"
			"      file if this is ordered vs. the reference.) If no --covStats or --regions option is specified the\n"
			"      output will only have aligned lengths. A value of '-' may be used to pipe this output to STDOUT.\n"
			"      Usage notes: The association to original read is only made by the line number, which directly\n"
			"      corresponds to that in the BAM file, or record output by 'samtools view -F 4 <BAM file>'.\n"
			"      For this reason using a multiple input BAM files is not recommended. (Pre-merge instead.)\n"
			"      Similarly, if input BAM read filtering options -L, -Q, -d or -u are employed (see below) then\n"
			"      the SAM read line should be created using equivalent samtools view options.\n"
			"  -R,--regions <file>       Input BED file for defining 'on-target' regions flagged in the BBC and\n"
			"      and other coverage files created. A value of '-' may be used to explicitly indicate no target\n"
			"      regions and may be used with an input BBC file to remove on-target information.\n"
			"  -S,--sumStats <file>      File name for reads summary statistics. Records total forward and reverse\n"
			"      reads for each contig/chromosome in the reference and the same for on-target reads if the\n"
			"      --regions (-R) option is provided. A value of '-' may be used to pipe this output to STDOUT.\n"
			"  -T,--readType <type>      Specifies the type of reads recorded in the input BAM file(s) and affects\n"
			"      the statistics recorded to the file created from using the --covStats (-C) option. Values are:\n"
			"    'covdepth' (default) => output region loci and coverage at depths - see --covDepths (-D) option.\n"
			"    'trgbases' => 'covdepth' output plus coverage statistics for forward/reverse base reads on target.\n"
			"      Statistics include the numbers of total bases covered, uncovered at 5' and 3' target ends, and\n"
			"      average base read depth for the individual targets. NOTE: Forward, reverse and average base reads\n"
			"      may be counted more than once for overlapping targets. (Using a merged BED file is recommended.)\n"
			"    'trgreads' => 'covdepth' output plus coverage statistics for F/R reads on target. Overlapping and\n"
			"      assigned read counts are provided, where reads are assigned to the target they overlap the most.\n"
			"      Counts for target end-to-end reads are also given, according to the --e2eGap (-E) option value.\n"
			"      This option requires input BAM files, since BBC files do not record base read origins.\n"
			"    'amplicon' => 'trgreads' output but reads are assigned primarily by there 5' read end vs. the target\n"
			"      end for the strand, assuming this where the PCR primer targeted. (Refer to --primerLength option.)\n"
			"    'AmpliSeq' => 'amplicon' output but with different default values for the --primerLength option (30),\n"
			"      to account for 5' primer digestion, and --e2eGap option (2).\n"
			"  -s,--samdepth             Output base coverage in samtools depth format to STDOUT. This option is\n"
			"      not compatible with BBC file creation options and assumes the --onTargetBases (-b) option if\n"
			"      the --regions (-R) option is also used. Note that bbctools has better performance than samtools\n"
			"      depth, especially at high coverage depths and/or in combination with the --regions (-R) option.\n"
			"  -i,--index                Create a BBC index (BCI) file in addition to a BBC file or from an input\n"
			"      BBC file (with a .bci in place of the .bbc file extension). This file allows for fast random\n"
			"      access in to a BBC file using bbctools view, similar to a BAI file for a BAM file access.\n"
			"  -c,--coarse               Create a coarse base coverage (CBC) file in addition to a BBC file or\n"
			"      from an input BBC file (with a .cbc in place of the .bbc file extension). This file allows for\n"
			"      rapid evaluation of summed coverage statistics over extended regions using bbctools view.\n\n"
			"Secondary file creation/output options:\n"
			"  -A,--annotationFields <list>  Output the specified fields from the --regions file directly to the\n"
			"      coverage statistics output file created from using the --covStats (-C) option, preceding the\n"
			"      calculated values. <list> is a comma-separated list of 0-based indexes for target fields to\n"
			"      report and may be negative to indicate an offset from the last field. Typically output field\n"
			"      names should also be specified using a ':' suffix to the indexes, e.g. '3:ampId,4:score'.\n"
			"  -D,--covDepths <list>     List of base coverage depths to output as extra fields to the coverage\n"
			"      statistics output file created from using the --covStats (-C) option. If not specified, a default\n"
			"      value of '20,100,500' is used, resulting in output fields 'cov20x', 'cov100x' and 'cov500x'. These\n"
			"      are the number of bases of the corresponding target (or contig) covered by at least 20, 100 and\n"
			"      500 reads respectively. A value of '' may be used to remove any coverage-at-depth statistics.\n"
			"  -E,--e2eGap <int>         Specifies the minimum distance the ends of an aligned read may fall short\n"
			"      of either end of the target region it is assigned to be considered an end-to-end read. This option\n"
			"      is only used with the --covStats (-C) option for --readType (-T) values 'trgreads', 'amplicon' or\n"
			"      'AmpliSeq' and may be overridden by the --minPcCov (-M) option (see below). Default value is 0.\n"
			"  -L,--minAlignLength <int> Minimum aligned read length filter for excluding reads when reading BAM files.\n"
			"      The default value of 0 means that no reads are excluded based on their alignment length.\n"
			"  -M,--minPcCov <percent>   Specifies the minimum percentage of the target that a read overlaps for it\n"
			"      to be counted as a 'covering' read. If the value is positive then the counts of covering reads per\n"
			"      target per strand will be output to the --covStats (-C) file as the 'fwd_cov' and 'rev_cov' fields\n"
			"      in preference to the end-to-end read counts ('fwd_e2e' and 'rev_e2e'). Default value is 0.\n"
			"  -P,--primerLength <int>   Specifies the assumed length of (digested) primer when assigning reads to\n"
			"      targets using --covStats (-C) for --readType 'amplicon' or 'AmpliSeq' options. Default value is 0.\n"
			"  -Q,--minMAPQ <int>        Minimum aligned read MAPQ value for excluding reads when reading BAM files.\n"
			"      The default value of 0 means that no reads are excluded based on SAM mapping quality score.\n"
			"  -a,--autoCreateBamIndex   Automatically create a BAM index (BAI) file if not located for (any of) the\n"
			"      BAM file(s) provided. Effective only if an on-target-only option is specified (-s, -b or -r).\n"
			"      Note: A run without a BAI file is typically faster than creating one on-the-fly and using it.\n"
			"  -b,--onTargetBases        Specifies to only output coverage for bases that overlap target regions in\n"
			"      the --regions (-R) file.\n"
			"  -d,--noDups               Specifies to ignore reads that are marked as duplicates in input BAM files.\n"
			"  -r,--onTargetReads        Specifies to only output coverage for reads that overlap target regions in\n"
			"      the --regions (-R) file. This option requires BAM input file(s).\n"
			"  -u,--unique               Specifies to ignore non-uniquely mapped reads from input BAM files.\n"
			"      This option is currently the same as using --minMAPQ 1.\n\n"
			"Examples:\n"
			"  1) bbctools create -ciB output -R input.bed input.bam\n"
			"       Create BBC format files, output.bbc, output.bci and output.cbc, indicating on-target base coverage\n"
			"       for regions in input.bed. BCI and CBC files are used for rapid coverage review with bbctools view.\n"
			"  2) bbctools create -C amp.cov.xls -T AmpliSeq -A 3:amp_id -R input.bed input.bam\n"
			"       Create read coverage summary file amp.cov.xls for AmpliSeq (insert) targets in input.bed, with the\n"
			"       4th field of input.bed brought over with the field name 'amp_id' (inserted after locus fields).\n"
			"  3) bbctools create -S chr.cov.xls -B output input.bbc\n"
			"       Create output.bbc for (non-targeted) base reads and contig reads file chr.cov.xls from BBC file.\n"
			"  4) bbctools create -sd -L 17 -Q 20 -R input.bed input.bam > sam.depth\n"
			"       Output (to sam.depth) total base reads in regions in input.bed, ignoring reads marked as duplicates,\n"
			"       or having less than 17 aligned bases or MAPQ values less than 20. As with the samtools depth command,\n"
			"       coverage is reported only for on-target bases, possibly including bases with 0 reads over deletions.\n"
			"  5) bbctools create -ciB output -R input.bed input.bbc\n"
			"       Create BBC format files, output.bbc, output.bci and output.cbc from existing BBC file input.bbc.\n"
			"       This might be to re-version the BBC file or re-defined the on-target bases.\n"
			"  6) bbctools create -C dpt.xls -D 1,100,1000 -R input.bed input.bbc\n"
			"       Create dpt.xls for coverage at 1x, 100x and 1000x read depths for targets in input.bed from BBC file.\n";
		} else {
			cout << "Options:\n"
			"  -B,--bbc <file prefix>    File name (prefix) for BBC file and other bbctools binary file creation.\n"
			"  -C,--covStats <file>      File name for coverage statistics output (text) file. (May require BAM input.)\n"
			"  -R,--regions <file>       Input BED file for defining 'on-target' regions in the output files.\n"
			"  -S,--sumStats <file>      File name for reads summary statistics (per contig). (Requires BAM input.)\n"
			"  -T,--readType <type>      Type of reads recorded in the input BAM file(s) for assessment with -C option.\n"
			"  -i,--index                Create a base coverage index (BCI) file. (May use BBC file input.)\n"
			"  -c,--coarse               Create a coarse base coverage (CBC) file. (May use BBC file input.)\n"
			"  -L,--minAlignLength <int> Minimum aligned read length filter for excluding reads when reading BAM files. (0)\n"
			"  -Q,--minMAPQ <int>        Minimum MAPQ value for excluding reads when reading BAM files. (0)\n"
			"  -d,--noDups               Ignore reads that are marked as duplicates in input BAM files.\n\n"
			"Type: 'bbctools create --help' for more options and detailed information on command arguments.\n";
		}
	} else if( subcmd == "report" ) {
		cout << "Usage:\n"
		"  bbctools report [options] <BBC file> [<window>]\n\n";
		if( verbose ) {
			cout << "Description:\n"
			"  Report some general statistics for base coverage from the given BBC file. This includes base strand bias\n"
			"  and (target) bases covered at specified read depths. Optionally a base coverage file may be output.\n"
			"  <window> is an optional reference locus, e.g. 'chr1', 'chr1:1000001' or 'chr1:1000001-1100001, or\n"
			"  trans-contig, e.g. 'chr1..chr2' or 'NM10016:2001..NM10085:1000'. Default locus is the whole reference.\n\n"
			"Options:\n"
			"  -g,--genomeReport         Report simplified base coverage statistics. Base coverage is labeled either\n"
			"      'Target' or 'Genome'. 'Target' refers to the combined focus of the <window>, --regions (-R) file and\n"
			"      -t options, if specified, with no additional statistics reported for BBC on-target base coverage.\n"
			"  -t,--reportOnTarget       Report on-target base read statistics for just bases marked as 'on-target' in the\n"
			"       BBC file that have at least one read (and within the target regions and/or range window if specified).\n"
			"  -C,--coverage <file>      Output base coverage to this file. This is a tab-delimited text file for non-zero\n"
			"      counts of bases covered at read depths, plus fields for the cumulative base counts at or above that\n"
			"      depth, the normalized read depths (divided by the mean base read depth) and the % cumulative base counts.\n"
			"  -D,--covDepths <list>     List of base coverage depths for percentage coverage at depths reported.\n"
			"      If not specified, a default value of '1,20,100,500' is employed.\n"
			"  -R,--regions <BED file>   Use this BED format file to specify target regions where coverage is viewed.\n"
			"      If <window> is also provided only (partial) regions within this window range are viewed.\n"
			"  -B,--strandBiasThreshold  Percentage of bases that have more than this fraction of either forward or\n"
			"      reverse reads for it to be counted as having strand bias. Bases are considered to have no strand bias\n"
			"      if they do not have enough reads enough reads. (See -M,--strandMinReads options.) Default 70.\n"
			"  -M,--statsMinReads        Minimum reads a single base must have before statistics such as strand bias are\n"
			"      assessed for that base position. (See -B,--strandBiasThreshold option.) Default 10.\n"
			"  -U,--uniformityThreshold  Percentage of the mean read depth at which percentage target cumulative\n"
			"      coverage is reported as the uniformity of base coverage statistics. Default 20.\n\n"
			"Examples:\n"
			"  1) bbctools report input.bbc\n"
			"       Output the default report summary for the contents of the input BBC file.\n"
			"  2) bbctools report -R input.bed input.bbc\n"
			"       Output the default report summary for the contents of the input BBC file over targets in input.bed.\n"
			"  3) bbctools report -B 60 -D 100,1000 -U 5 input.bbc chrM\n"
			"       Output the report summary for the input BBC file for just contig 'chrM' but % base coverage reported\n"
			"       for read depths of 100+ and 1000+, the uniformity threshold at 5% MRD and a 60% strand bias threshold.\n"
			"  4) bbctools report -t -C doc.xls -R input.bed input.bbc chr1\n"
			"       Output the report summary for the contents of the input BBC file for targets in input.bed on contig\n"
			"       'chr1' only and create the depth of coverage table to doc.xls for bases in these regions. Note that\n"
			"       the -t option further restricts to bases marked on-target in the BBC file with at least one read.\n";
		} else {
			cout << "Options:\n"
			"  -g,--genomeReport         Report simplified base coverage statistics.\n"
			"  -C,--coverage <file>      Output base coverage (counts of base read depths) to this file.\n"
			"  -D,--covDepths <list>     List of base coverage depths for percentage coverage at depths reported.\n"
			"  -R,--regions <BED file>   BED format file to specify target regions where coverage is viewed.\n\n"
			"Type: 'bbctools report --help' for more options and detailed information on command arguments.\n";
		}
	} else if( subcmd == "view" ) {
		cout << "Usage:\n"
		"  bbctools view [options] <BBC file> [<window>]\n\n";
		if( verbose ) {
			cout << "Description:\n"
			"  Review base coverage or histogram of (binned) base coverage over a specified range or target regions.\n"
			"  The default output is tab-delimited text starting with the contig/chromosome name, followed by the\n"
			"  base position, 1 or 0 to indicate that the base was on-target and total forward and reverse reads.\n"
			"  Binned coverage views output total base read depth over contiguous sub-regions of the reference, with\n"
			"  contig(s) start and end positions for each region followed by separate total and on-target read counts\n"
			"  for forward and reverse reference strands. Options specify the binning method and output formats.\n"
			"  <window> is an optional reference locus, e.g. 'chr1', 'chr1:1000001' or 'chr1:1000001-1100001, or\n"
			"  trans-contig, e.g. 'chr1..chr2' or 'NM10016:2001..NM10085:1000'. Default locus is the whole reference.\n\n"
			"Region coalescence (binning) options:\n"
			"  -R,--regions <BED file>   Use this BED format file to specify target regions where coverage is viewed.\n"
			"      If <window> is also provided only (partial) regions within this window range are viewed.\n"
			"      NOTE: The regions file may alternatively be used for annotation only - refer to -A option below.\n"
			"  -i,--isolateRegions       Output summed coverage to range bins such that each contig will be\n"
			"      represented in a separate number of region bins, the number being apportioned by the relative\n"
			"      sizes of the contigs using --numbins or a fixed size using --binsize. The value for --numbins\n"
			"      will be reset to the number of contigs if less. If neither --numbins or --binsize are\n"
			"      specified (>0), the default behavior is to show summed coverage by contig. If the --region\n"
			"      option is specified the description above applies to isolated regions rather than contigs.\n"
			"      If <window> is specified the regions or contigs are restricted to those within this range.\n"
			"      NOTE: If this option is not provided the regions or contigs within the window are treated as\n"
			"      a contiguous sequence of bases, i.e. individual binned regions may span regions or contigs.\n"
			"  -B,--binsize <integer>    Output summed coverage in regions of this size spread evenly across the\n"
			"      whole regions/window range such that only the last region (bin) may cover less than this\n"
			"      number of bases. Using the default value (0) with --numbins also at the default value (0) is\n"
			"      equivalent to using 1 except that only a single base coverage output format is used, rather\n"
			"      than a ranged output format, and some output formating options are not be compatible.\n"
			"      Using non-zero values for both --binsize and --numbins is disallowed.\n"
			"  -N,--numbins <integer>    Output summed coverage over this number of regions of (near) equal size\n"
			"      across the whole regions/window range. If the mean bin size is not an integer, the fractional\n"
			"      part is spread evenly across the range such that some bin sizes contain 1 more base.\n"
			"      Using non-zero values for both --binsize and --numbins is disallowed.\n"
			"  -S,--startBin <integer>   Output summed coverage starting at this region bin of the total bins\n"
			"      that would be output given the other binning options (-B, -N or -i). The value may be\n"
			"      negative, in which case it is an offset from what would be the last binned region output.\n"
			"      The default value (0) is entirely equivalent to 1, i.e. output from the first region bin.\n"
			"  -E,--endBin <integer>     Output summed coverage ending at this region bin of the total bins\n"
			"      that would be output given the other binning options (-B, -N or -i). The value may be\n"
			"      negative, in which case it is an offset is from what would be the last binned region output.\n"
			"      The default value (0) indicates to output up to and including the last region bin.\n\n"
			"Output format options:\n"
			"  -A,--annotationFields <list>  Output the specified fields from the --regions file for targets that\n"
			"      overlap output loci (bins). If a output region overlaps multiple targets the values are comma\n"
			"      separated or <first>,...(X)...,<last> for more than 3 values (X>1). <list> is a comma-separated\n"
			"      list of 0-based indexes for target fields to report and may be negative to indicate an offset\n"
			"      from the last field, e.g. '3' for the 4th (target name) field. NOTE: When the -A options is used\n"
			"      the regions file is used for annotation only and no longer defines coverage target regions.\n"
			"  -H,--headerLine <text>    Output this text as the first line of output. The default value ('')\n"
			"      means no header line is output. Comma characters in the text are automatically converted to\n"
			"      tab characters so as to create column titles to align with the field values output. This may\n"
			"      be prevented by using a double forward slash escape sequence, e.g. 'x\\\\,y'.\n"
			"  -b,--bedCoordinates       Output start position coordinates as 0-based, as used in BED files.\n"
			"  -c,--compactCoverage      Output contig/chromosome names on a separate line once rather than for\n"
			"      every base position. (Contig lines start with '>'.) This option is only available for the\n"
			"      (default) base coverage view (i.e. where no -B, -N or -i options are also provided).\n"
			"  -l,--lociOnly             Output only the region/base loci, i.e. no coverage data.\n"
			"  -n,--omitContigNames      Omit the contig/chromosome names from the output.\n"
			"  -p,--omitContigPositions  Omit the contig/chromosome position coordinates from the output.\n"
			"  -s,--sumStrandCoverage    Output the sums of coverage on both strands rather than separate\n"
			"      forward and reverse values; for both total and on-target coverage.\n"
			"  -t,--omitTargetCoverage   Omit the on-target flag or on-target read fields from the output.\n"
			"  -z,--includeZeroCoverage  Include output for bases with 0 total coverage. This option is only\n"
			"      available for base coverage view (i.e. where no -B, -N or -i options are also provided).\n\n"
			"Examples:\n"
			"  1) bbctools view input.bbc\n"
			"       Review individual base coverage, in format <chr>TAB<pos>TAB<on-trg>TAB<fwd-rds>TAB<rev-rds>.\n"
			"  2) bbctools view -B 1000000 input.bbc chr1\n"
			"       Report total on and off target base reads per strand across contig 'chr1' over 1M base sections.\n"
			"  3) bbctools view -N 1 -R input.bed input.bbc chr1\n"
			"       Report total sum of on and off target base reads over all regions in input.bed on contig 'chr1'.\n"
			"  4) bbctools view -N 600 -pst -H 'contig,bases' input.bbc\n"
			"       Report total base reads across entire reference in 600 contiguous sections. The -p, -s and -t\n"
			"       options result in only contig names and total reads being reported, with a suitable header line.\n"
			"  5) bbctools view -i -H 'chr,pos,f_r,r_r,f_tr,r_tr' -R input.bed input.bbc\n"
			"       Report total on and off target base reads for each region in input.bed, with given column headers.\n"
			"  6) bbctools view -i -H 'chr,pos,f_r,r_r,f_tr,r_tr,rid' -A 3 -R input.bed input.bbc\n"
			"       Report total on and off target base reads per contig with given column headers, including 'rid'\n"
			"       for the 4th field of input.bed. The -A option overrides use of input.bed to define target regions.\n";
		} else {
			cout << "Options:\n"
			"  -R,--regions <BED file>   BED file to specify target coverage or annotation (with -A) regions.\n"
			"  -i,--isolateRegions       Isolate each contig or region (with -R) for histogram view bins.\n"
			"  -B,--binsize <integer>    Use this fixed bin size for coverage (histogram) view. (0)\n"
			"  -N,--numbins <integer>    Use this total number of bins for coverage (histogram) view. (0)\n"
			"  -A,--annotationFields <list>  Add specified fields from the -R file to (histogram) bins output.\n"
			"  -H,--headerLine <text>    Column names for (tab-separated) output header (comma separated). ('' => none)\n"
			"  -s,--sumStrandCoverage    Output the sums of coverage on both strands.\n\n"
			"Type: 'bbctools view --help' for more options and detailed information on command arguments.\n";
		}
	} else if( subcmd == "version" ) {
		cout << "Usage:\n"
		"  bbctools version [options] [<BBC|BCI|CBC file>]\n\n"
		"Description:\n"
		"  Returns the version number of the current bbctools executable or details of bbctools binary file.\n\n"
		"Options:\n"
		"  -b,--brief  Just return version number with no extra formating or information.\n";
	} else {
		cerr << "Error: Invalid bbctools sub-command '" << subcmd << "'.\n\n"
		"Use 'bbctools' with no arguments to print a summary of usage, or use 'bbctools <subcmd> -h [--help]'\n"
		"for [detailed] information on a particular bbctools sub-command and its specific options.\n";
	}
}

#endif /* BBCUSAGE_H_ */
