diff -crB orig/armadillo-2.4.0/CMakeLists.txt armadillo-2.4.0/CMakeLists.txt
*** orig/armadillo-2.4.0/CMakeLists.txt	2011-12-04 19:04:15.000000000 -0800
--- armadillo-2.4.0/CMakeLists.txt	2012-03-28 08:55:18.927801000 -0700
***************
*** 70,79 ****
    
  else()
    set(ARMA_OS unix)
!   
!   include(ARMA_FindMKL)
!   include(ARMA_FindACMLMP)
!   include(ARMA_FindACML)
    
    message(STATUS "MKL_FOUND     = ${MKL_FOUND}")
    message(STATUS "ACMLMP_FOUND  = ${ACMLMP_FOUND}")
--- 70,80 ----
    
  else()
    set(ARMA_OS unix)
! 
! # Uncomment these lines to use intel math libraries.  
! #  include(ARMA_FindMKL)
! #  include(ARMA_FindACMLMP)
! #  include(ARMA_FindACML)
    
    message(STATUS "MKL_FOUND     = ${MKL_FOUND}")
    message(STATUS "ACMLMP_FOUND  = ${ACMLMP_FOUND}")
Only in armadillo-2.4.0: CMakeLists.txt~
diff -crB orig/armadillo-2.4.0/include/armadillo_bits/config.hpp armadillo-2.4.0/include/armadillo_bits/config.hpp
*** orig/armadillo-2.4.0/include/armadillo_bits/config.hpp	2011-11-18 10:27:31.000000000 -0800
--- armadillo-2.4.0/include/armadillo_bits/config.hpp	2012-03-28 08:54:53.683267000 -0700
***************
*** 21,34 ****
  //// This will enable additional features, such as use of initialiser lists
  
  #if !defined(ARMA_USE_LAPACK)
! // #define ARMA_USE_LAPACK
  //// Uncomment the above line if you have LAPACK or a fast replacement for LAPACK,
  //// such as Intel's MKL, AMD's ACML, or the Accelerate framework.
  //// LAPACK is required for matrix decompositions (eg. SVD) and matrix inverse.
  #endif
  
  #if !defined(ARMA_USE_BLAS)
! // #define ARMA_USE_BLAS
  //// Uncomment the above line if you have BLAS or a fast replacement for BLAS,
  //// such as GotoBLAS, Intel's MKL, AMD's ACML, or the Accelerate framework.
  //// BLAS is used for matrix multiplication.
--- 21,34 ----
  //// This will enable additional features, such as use of initialiser lists
  
  #if !defined(ARMA_USE_LAPACK)
! #define ARMA_USE_LAPACK
  //// Uncomment the above line if you have LAPACK or a fast replacement for LAPACK,
  //// such as Intel's MKL, AMD's ACML, or the Accelerate framework.
  //// LAPACK is required for matrix decompositions (eg. SVD) and matrix inverse.
  #endif
  
  #if !defined(ARMA_USE_BLAS)
! #define ARMA_USE_BLAS
  //// Uncomment the above line if you have BLAS or a fast replacement for BLAS,
  //// such as GotoBLAS, Intel's MKL, AMD's ACML, or the Accelerate framework.
  //// BLAS is used for matrix multiplication.
***************
*** 59,66 ****
  // #define ARMA_USE_TBB_ALLOC
  //// Uncomment the above line if you want to use Intel TBB scalable_malloc() and scalable_free() instead of standard new[] and delete[]
  
! // #define ARMA_USE_ATLAS
! // #define ARMA_ATLAS_INCLUDE_DIR /usr/include/
  //// If you're using ATLAS and the compiler can't find cblas.h and/or clapack.h
  //// uncomment the above define and specify the appropriate include directory.
  //// Make sure the directory has a trailing /
--- 59,66 ----
  // #define ARMA_USE_TBB_ALLOC
  //// Uncomment the above line if you want to use Intel TBB scalable_malloc() and scalable_free() instead of standard new[] and delete[]
  
! #define ARMA_USE_ATLAS
! #define ARMA_ATLAS_INCLUDE_DIR /usr/include/
  //// If you're using ATLAS and the compiler can't find cblas.h and/or clapack.h
  //// uncomment the above define and specify the appropriate include directory.
  //// Make sure the directory has a trailing /
***************
*** 72,84 ****
  #define ARMA_PRINT_LOGIC_ERRORS
  #define ARMA_PRINT_RUNTIME_ERRORS
  
! // #define ARMA_HAVE_STD_ISFINITE
! // #define ARMA_HAVE_STD_ISINF
! // #define ARMA_HAVE_STD_ISNAN
! // #define ARMA_HAVE_STD_SNPRINTF
  
! // #define ARMA_HAVE_LOG1P
! // #define ARMA_HAVE_GETTIMEOFDAY
  
  // #define ARMA_EXTRA_DEBUG
  // #define ARMA_NO_DEBUG
--- 72,84 ----
  #define ARMA_PRINT_LOGIC_ERRORS
  #define ARMA_PRINT_RUNTIME_ERRORS
  
! #define ARMA_HAVE_STD_ISFINITE
! #define ARMA_HAVE_STD_ISINF
! #define ARMA_HAVE_STD_ISNAN
! #define ARMA_HAVE_STD_SNPRINTF
  
! #define ARMA_HAVE_LOG1P
! #define ARMA_HAVE_GETTIMEOFDAY
  
  // #define ARMA_EXTRA_DEBUG
  // #define ARMA_NO_DEBUG
diff -crB orig/armadillo-2.4.0/include/armadillo_bits/fn_norm.hpp armadillo-2.4.0/include/armadillo_bits/fn_norm.hpp
*** orig/armadillo-2.4.0/include/armadillo_bits/fn_norm.hpp	2011-11-01 08:23:38.000000000 -0700
--- armadillo-2.4.0/include/armadillo_bits/fn_norm.hpp	2012-03-28 08:54:53.691199000 -0700
***************
*** 433,438 ****
--- 433,439 ----
          return T(0);
        }
      }
+   return T(0);
    }
  
  
diff -crB orig/armadillo-2.4.0/include/armadillo_bits/unwrap.hpp armadillo-2.4.0/include/armadillo_bits/unwrap.hpp
*** orig/armadillo-2.4.0/include/armadillo_bits/unwrap.hpp	2011-07-09 00:41:15.000000000 -0700
--- armadillo-2.4.0/include/armadillo_bits/unwrap.hpp	2012-03-28 08:54:53.696830000 -0700
***************
*** 1205,1210 ****
--- 1205,1211 ----
      , M  (A.m)
      {
      arma_extra_debug_sigprint();
+     arma_ignore(B);
      }
    
    inline
