--- CMakeLists.txt	2013-03-25 00:54:00.000000000 -0400
+++ CMakeLists.txt	2013-04-05 15:31:12.873288265 -0400
@@ -31,14 +31,25 @@
 #CMAKE_REQUIRED_DEFINITIONS = list of macros to define (-DFOO=bar)
 #CMAKE_REQUIRED_LIBRARIES = list of libraries to link
 
-
-set(ARMA_USE_LAPACK  false)
-set(ARMA_USE_BLAS    false)
-set(ARMA_USE_ATLAS   false)
-set(ARMA_USE_BOOST   false)
-set(ARMA_USE_HDF5    false)
-set(ARMA_USE_WRAPPER true )  # set this to false if you prefer to directly link with LAPACK and/or BLAS (eg. -llapack -lblas) instead of -larmadillo
-
+option(ARMA_USE_LAPACK "Use LAPACK libraries" OFF)
+mark_as_advanced(ARMA_USE_LAPACK)
+option(ARMA_USE_BLAS "Use BLAS libraries" OFF)
+mark_as_advanced(ARMA_USE_BLAS)
+option(ARMA_USE_ATLAS "Use ATLAS libraries" OFF)
+mark_as_advanced(ARMA_USE_ATLAS)
+option(ARMA_USE_BOOST "Use BOOST libraries" OFF)
+mark_as_advanced(ARMA_USE_BOOST)
+option(ARMA_USE_HDF5 "Use HDF5 libraries" OFF)
+mark_as_advanced(ARMA_USE_HDF5)
+# set this to false if you prefer to directly link with LAPACK and/or BLAS (eg. -llapack -lblas) instead of -larmadillo
+option(ARMA_USE_WRAPPER "Use WRAPPER libraries" OFF)
+mark_as_advanced(ARMA_USE_WRAPPER)
+option(ARMA_USE_MKL "Use MKL libraries" OFF)
+mark_as_advanced(ARMA_USE_MKL)
+option(ARMA_USE_ACMLMP "Use ACMLMP libraries" OFF)
+mark_as_advanced(ARMA_USE_ACMLMP)
+option(ARMA_USE_ACML "Use ACML libraries" OFF)
+mark_as_advanced(ARMA_USE_ACML)
 
 if(WIN32)
   message(STATUS "")
@@ -68,15 +79,22 @@
   
 else()
   set(ARMA_OS unix)
-  
-  include(ARMA_FindMKL)
-  include(ARMA_FindACMLMP)
-  include(ARMA_FindACML)
-  
-  message(STATUS "MKL_FOUND     = ${MKL_FOUND}")
-  message(STATUS "ACMLMP_FOUND  = ${ACMLMP_FOUND}")
-  message(STATUS "ACML_FOUND    = ${ACML_FOUND}")
-  
+
+  if(ARMA_USE_MKL)
+    include(ARMA_FindMKL)
+    message(STATUS "MKL_FOUND     = ${MKL_FOUND}")
+  endif()
+
+  if(ARMA_USE_ACMLMP)
+    include(ARMA_FindACMLMP)
+    message(STATUS "ACMLMP_FOUND  = ${ACMLMP_FOUND}")
+  endif()
+
+  if(ARMA_USE_ACML)  
+    include(ARMA_FindACML)
+    message(STATUS "ACML_FOUND    = ${ACML_FOUND}")
+  endif()
+
   if(MKL_FOUND OR ACMLMP_FOUND OR ACML_FOUND)
     
     set(ARMA_USE_BLAS true)
--- include/armadillo_bits/fn_norm.hpp	2013-02-21 20:00:53.000000000 -0500
+++ include/armadillo_bits/fn_norm.hpp	2013-04-05 16:02:03.293356218 -0400
@@ -530,6 +530,7 @@
         return T(0);
       }
     }
+  return T(0);
   }
 
 
--- include/armadillo_bits/unwrap.hpp	2013-02-28 02:58:00.000000000 -0500
+++ include/armadillo_bits/unwrap.hpp	2013-04-05 16:03:44.301359941 -0400
@@ -1601,6 +1601,7 @@
     //, ref( A )
     {
     arma_extra_debug_sigprint();
+    arma_ignore(B);
     }
   
   arma_hot arma_inline eT get_val() const { return eT(1); }
