--- a/include/host_config.h	2013-07-19 09:23:44.000000000 -0400
+++ b/include/host_config.h	2013-10-10 10:59:53.000000000 -0400
@@ -69,7 +69,7 @@
 
 #if !(__INTEL_COMPILER == 9999 && __INTEL_COMPILER_BUILD_DATE == 20110811) || !defined(__GNUC__) || !defined(__LP64__)
 
-#error -- unsupported ICC configuration! Only ICC 12.1 on Linux x86_64 is supported!
+//#error -- unsupported ICC configuration! Only ICC 12.1 on Linux x86_64 is supported!
 
 #endif /* !(__INTEL_COMPILER == 9999 && __INTEL_COMPILER_BUILD_DATE == 20110811) || !__GNUC__ || !__LP64__ */
 
@@ -79,7 +79,7 @@
 
 #if __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ > 8)
 
-#error -- unsupported GNU version! gcc 4.9 and up are not supported!
+//#error -- unsupported GNU version! gcc 4.9 and up are not supported!
 
 #endif /* __GNUC__> 4 || (__GNUC__ == 4 && __GNUC_MINOR__ > 8) */
 
